starter/source/materials/fail/fractal/random_walk_dmg.F90

```diff
@@ -56,248 +56,254 @@ subroutine random_walk_dmg(fractal,fail ,                           &
+              end if
             end if
-            if (target_gsh3_id > 0) then
-              igr3n_target = ngr2usrn(target_gsh3_id,igrsh3n,ngrsh3n,nb_sh3)
+          end do
+        end do
+        do i=1,nb_sh3
+          iel = igrsh3n(igr3n_target)%entity(i)
+          do j = 1,fractal%nelem
+            if (fractal%random_walk(j)%nix == 3) then
+              if (fractal%random_walk(j)%elnum == iel) then
+                fractal%random_walk(j)%damage = dmg
+                if (debug==1) print*,'    ',fractal%random_walk(j)%id
+              end if
             end if
-            !  initialize damage in target elements
-            do i=1,nb_sh4
-              iel = igrsh4n(igr4n_target)%entity(i)
-              do j = 1,fractal%nelem
-                if (fractal%random_walk(j)%nix == 4) then
-                  if (fractal%random_walk(j)%elnum == iel) then
-                    fractal%random_walk(j)%damage = dmg
-                    if (debug==1) print*,'    ',fractal%random_walk(j)%id
-                  end if
-                end if
-              end do
-            end do
-            do i=1,nb_sh3
-              iel = igrsh3n(igr3n_target)%entity(i)
-              do j = 1,fractal%nelem
-                if (fractal%random_walk(j)%nix == 3) then
-                  if (fractal%random_walk(j)%elnum == iel) then
-                    fractal%random_walk(j)%damage = dmg
-                    if (debug==1) print*,'    ',fractal%random_walk(j)%id
-                  end if
-                end if
-              end do
-            end do
-          endif       ! random target
+          end do
+        end do
+      endif       ! random target
 !------------------------------------------------------------------------------------
 !     fractal random walks from starting elements to targets
 !------------------------------------------------------------------------------------
+      if (debug==1) then
+        print*,' '
+        print*,' begin random walk algorithm'
+      end if
+      !-------------------------------------
+!
+      do iwalk=1,n_rwalk
+        stop_walk = 0
+        call random_number(random_value)
+        iel = ceiling(random_value * nstart)
+        current = start_group(iel)
+        ! skip already damaged elements => choose another starting point
+        if (fractal%random_walk(current)%damage == zero) then
           if (debug==1) then
             print*,' '
-            print*,' begin random walk algorithm'
+            print*,' drunken walker number',iwalk,',start with',fractal%random_walk(current)%id
           end if
-          !-------------------------------------
-!
-          do iwalk=1,n_rwalk
-            stop_walk = 0
+          !------------------------------------------------
+          ! start random walk from current starting point
+          !------------------------------------------------
+          prv = 0
+          do while (stop_walk == 0)
```

Comment:
The loop may be infinite. It is worth adding a maximum number of iterations and stopping with an error when it's reached.